circuits/cpp/src/aztec3/circuits/kernel/private/common.cpp

```diff
@@ -44,6 +46,30 @@ void common_validate_call_stack(DummyComposer& composer, PrivateCallData<NT> con
     }
 }
 
+void common_validate_read_requests(DummyComposer& composer, PrivateCallData<NT> const& private_call)
```

Comment:
@dbanks12 We validate that all read request belong to the same tree as I understand. Is there no check that the root is actually the genuine one?

Reply:
Good question! I think it will need to be checked against the historic root in `public_inputs.constants`, or maybe it actually determines that `public_inputs.constants`'s data root which is then validated in later circuits (rollup)

Reply:
My TODO at the bottom of this function is meant to capture this as well as one of my recent comments in the [discourse thread](https://discourse.aztec.network/t/spending-notes-which-havent-yet-been-inserted/180/10?u=david-banks).

Reply:
> According to [Mike's post on Discourse](https://discourse.aztec.network/t/spending-notes-which-havent-yet-been-inserted/180#changes-to-noir-circuit-logic-5)
we should set it to:
private_kernel_public_inputs.constant_data.old_tree_roots.private_data_tree_root

I missed this comment. I think you're right, but unsure.... It might be the opposite.... We might use it to initialize that constant root for the first time, and then never let it change and enforce that all read-requests use that same root. Then that constant root (since it is a public input to the private kernel), can be validated by the rollup circuits.

Reply:
https://discourse.aztec.network/t/proposal-spending-notes-which-havent-yet-been-inserted/180/14